### PR TITLE
#26545 Fix: same .tsv file, get different data-frame structure using engine 'python' and 'c' 

### LINKED ISSUE #26545: same .tsv file, get different data-frame structure using engine 'python' and 'c'
In my Mac, I have a Tab-Separated values file with encoding UTF-8, and the version of pandas is 
0.24.2.
[74872_zh_CN_UI.txt](https://github.com/pandas-dev/pandas/files/3227255/74872_zh_CN_UI.txt)

when I use read_csv function with engine 'python' like this:
`b = pd.read_csv("/Users/GHIBLI/Documents/vmware-L10n/bert/zh_CN/74872_zh_CN_UI.tsv",
                engine="python", delimiter="\t")`
` print(b.shape)`
I got (8, 1)

if with default engine:
`b = pd.read_csv("/Users/GHIBLI/Documents/vmware-L10n/bert/zh_CN/74872_zh_CN_UI.tsv",
                delimiter="\t")`
` print(b.shape)`
I got (8,22)

In contrast to 'C' engine, this 'python' engine seems that is not as simple as  just 'feature-complete' I think.

### PR DESCRIPTION
**Pull request by luckydenis**
- [x] closes #26545
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

BugFix:
When using `engine='python'`, columns were handled incorrectly if the first header had in the bom.

Bug:
```
In [1]: import pandas as pd                                                             

In [2]: pd.read_csv('test.txt', engine='python', delimiter='\t')              
Out[2]: 
Empty DataFrame
Columns: [Project ID]
Index: []

In [3]: pd.read_csv('test.txt', engine='python', delimiter='\t').shape        
Out[3]: (0, 1)

In [4]: pd.read_csv('test.txt', delimiter='\t')                               
Out[4]: 
Empty DataFrame
Columns: [Project ID, Project Name, Product Name]
Index: []

In [5]: pd.read_csv('test, delimiter='\t').shape                         
Out[5]: (0, 3)
```
Fix:

```
In [1]: import pandas as pd                                                             

In [2]: pd.read_csv('test.txt', engine='python', delimiter='\t')              
Out[2]: 
Empty DataFrame
Columns: [Project ID, Project Name, Product Name]
Index: []

In [3]: pd.read_csv('test.txt', engine='python', delimiter='\t').shape        
Out[3]: (0, 3)

In [4]: pd.read_csv('test.txt', delimiter='\t')                               
Out[4]: 
Empty DataFrame
Columns: [Project ID, Project Name, Product Name]
Index: []

In [5]: pd.read_csv('test, delimiter='\t').shape                         
Out[5]: (0, 3)
```

### CODE REVIEW

File: pandas/tests/io/parser/test_check_for_bom.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import pytest
+
+from pandas import read_csv
+
+
+def test_first_row_bom():
```
Comment (WillAyd) [inline pandas/tests/io/parser/test_check_for_bom.py]: Can you move this to `parser/test_common.py` and use the `all_parsers` fixture?

File: pandas/tests/io/parser/data/bom_first_line.txt [hunk 2]
```diff
@@ -0,0 +1,9 @@
+﻿"Project ID"	"Project Name"	"Product Name"	"Content Type"	"Creation Date"	"Translator"	"Reviewer"	"Lanugage"	"File Name"	"Segment Sequence"	"Word Count"	"TM Score"	"Match Type"	"AT Type"	"MT Engine"	"MT Configuration Name"	"PE Distance"	"Length After Automatic Translation"	"Change in %"	"Source"	"After Automatic Translation"	"Target"
```
Comment (WillAyd) [inline pandas/tests/io/parser/data/bom_first_line.txt]: Do we need an entire file for this or can you just construct a StringIO object directly? Minimal tests would be preferable

**Review by WillAyd — CHANGES_REQUESTED** [review]

**Comment by WillAyd** [comment]
This closes an issue right? If so can you update the OP to reflect that

File: pandas/tests/io/parser/test_check_for_bom.py (hunk 1 above)
Comment (LuckyDenis) [reply pandas/tests/io/parser/test_check_for_bom.py]: Yes, certainly. That makes sense.

File: pandas/tests/io/parser/data/bom_first_line.txt (hunk 2 above)
Comment (LuckyDenis) [reply pandas/tests/io/parser/data/bom_first_line.txt]: Yes, I can repeat the error using StringIO.

**Comment by LuckyDenis** [comment]
@WillAyd, I made a correction on your comments. Look please)

File: pandas/tests/io/parser/test_common.py [hunk 3]
```diff
@@ -1927,3 +1927,17 @@ def test_read_table_deprecated(all_parsers):
...
         tm.assert_frame_equal(result, expected)
+
+
+def test_first_row_bom_python(all_parsers):
+    parser = all_parsers
+    data = """
+    \ufeff"Head1"	"Head2"	"Head3"
+    """
+
+    assert parser.read_csv(StringIO(data),
```
Comment (WillAyd) [inline pandas/tests/io/parser/test_common.py]: Can you construct the actual DataFrame that should be represented and do `tm.assert_frame_equal(result, expected)`? Should see that idiom in other tests in this module if you need some examples to look at

File: pandas/io/parsers.py [hunk 4]
```diff
@@ -2755,23 +2755,24 @@ def _check_for_bom(self, first_row):
         if first_elt != _BOM:
             return first_row
 
-        first_row = first_row[0]
+        first_row_bom = first_row[0]
```
Comment (WillAyd) [inline pandas/io/parsers.py]: Hmm so the assumption here is that if there is a BOM the length of the first parsed element will be greater than 1, correct? Not sure if that's right or wrong just want to verify

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/tests/io/parser/test_common.py (hunk 3 above)
Comment (LuckyDenis) [reply pandas/tests/io/parser/test_common.py]: Yes, I saw how it was done above. But the problem was in the correct calculation of columns, DataFrame was built correctly, if the check will be in the form of `tm.assert_frame_equal(result.shape, expected.shape)` will be fine?

File: pandas/io/parsers.py (hunk 4 above)
Comment (LuckyDenis) [reply pandas/io/parsers.py]: Yes, there is an assumption that the first element will have a BOM, I do not know whether it is correct or not. But initially the error was in the other. We got the list. Mutated it into one element. Cleared this item from the BOM and given that this element is wrapped in a list.

File: pandas/tests/io/parser/test_common.py (hunk 3 above)
Comment (LuckyDenis) [reply pandas/tests/io/parser/test_common.py]: If I did this, would it be okay?

```
    assert result_c.shape == result_python.shape
    tm.assert_frame_equal(result_c, expected)
    tm.assert_frame_equal(result_python, expected)
```

Comment (WillAyd) [reply pandas/tests/io/parser/test_common.py]:
```python
result = parser.read_csv(StringIO(data), delimiter='\t')
expected = DataFrame(columns=["Head1", "Head2", "Head3"])
tm.assert_frame_equal(result, expected)
```

I think should work. Note that the parametrization for `all_parsers` makes it so you don't need to specify the engine as an argument

**Review by WillAyd — CHANGES_REQUESTED** [review]
Thanks for the updates - can you also add a whatsnew note for 0.25?

cc @gfyoung if you care to take a look

**Comment by WillAyd** [comment]
@LuckyDenis could you also check what `test_utf8_bom` is doing in the test module? It looks to cover the same intention as test added so want to make sure we understand the difference and clarify accordingly

**Comment by LuckyDenis** [comment]
In this context, I'm not sure that they are different, but my PR cures this error, which `tm` method should be used to check it?

```
In [1]: import pandas as pd                                                             

In [2]: pd.read_csv('test.txt', engine='python', delimiter='\t')              
Out[2]: 
Empty DataFrame
Columns: [Project ID]
Index: []

In [3]: pd.read_csv('test.txt', engine='python', delimiter='\t').shape        
Out[3]: (0, 1)

In [4]: pd.read_csv('test.txt', delimiter='\t')                               
Out[4]: 
Empty DataFrame
Columns: [Project ID, Project Name, Product Name]
Index: []

In [5]: pd.read_csv('test, delimiter='\t').shape                         
Out[5]: (0, 3)

```

File: pandas/tests/io/parser/test_common.py [hunk 5]
```diff
@@ -1927,3 +1927,13 @@ def test_read_table_deprecated(all_parsers):
...
+
+
+def test_first_row_bom(all_parsers):
+    # see gh-26545
+    parser = all_parsers
+    data = '''\ufeff"Head1"	"Head2"	"Head3"'''
+
+    result = parser.read_csv(StringIO(data), delimiter='\t')
+    expected = DataFrame(columns=["Head1", "Head2", "Head3"])
+    tm.assert_frame_equal(result, expected)
```
Comment (gfyoung) [inline pandas/tests/io/parser/test_common.py]: This test is fine to add.  `tm.assert_frame_equal` does the trick.

Comment (LuckyDenis) [reply pandas/tests/io/parser/test_common.py]: > @LuckyDenis could you also check what `test_utf8_bom` is doing in the test module? It looks to cover the same intention as test added so want to make sure we understand the difference and clarify accordingly

@WillAyd, I looked at the difference between the tests. In the `test_utf8_bom`, the dataset only checks that it clears the first header from the bom and does not reveal the bug I fixed. My test checks just processing headers and that you grow them correctly.

Comment (WillAyd) [reply pandas/tests/io/parser/test_common.py]: Awesome thanks for verifying that!

**Review by WillAyd — CHANGES_REQUESTED** [review]
I think this is good if you can add a whatsnew for 0.25 then lgtm!

**Comment by jreback** [comment]
@gfyoung ok with this?

**Comment by gfyoung** [comment]
@jreback : Looks fine to me, just need the `whatsnew` as @WillAyd said

**Comment by LuckyDenis** [comment]
Yeah, I'll add a description in the `whatsnew`, just haven't had time.

**Comment by LuckyDenis** [comment]
@jreback, @WillAyd, @gfyoung  I added, look please.

File: doc/source/whatsnew/v0.25.0.rst [hunk 6]
```diff
@@ -664,6 +664,7 @@ I/O
 - Bug in :func:`read_json` where date strings with ``Z`` were not converted to a UTC timezone (:issue:`26168`)
 - Added ``cache_dates=True`` parameter to :meth:`read_csv`, which allows to cache unique dates when they are parsed (:issue:`25990`)
 - :meth:`DataFrame.to_excel` now raises a ``ValueError`` when the caller's dimensions exceed the limitations of Excel (:issue:`26051`)
+- Fixed bug in :func:`pandas.read_csv` - when using engine='python', columns were handled incorrectly if the first header had in the bom (:issue:`26545`)
```
Comment (gfyoung) [inline doc/source/whatsnew/v0.25.0.rst]: "if the first header had in the bom" ? - can you reword this?

Comment (WillAyd) [reply doc/source/whatsnew/v0.25.0.rst]:
```suggestion
- Fixed bug in :func:`pandas.read_csv` - where a BOM would result in incorrect parsing using engine='python' (:issue:`26545`)
```

**Review by WillAyd — APPROVED** [review]
lgtm @gfyoung

**Review by gfyoung — APPROVED** [review]
@jreback if we have anything else

**Comment by jreback** [comment]
lgtm. @LuckyDenis if you can merge master and ping on green to resolve the conflict.

**Comment by jreback** [comment]
thanks @LuckyDenis

**Pull request merged**